---
title: "Class 15"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Transcriptomics and analysis of RNA-Seq Data
# 1: Bioconductor and DESeq2 Setup
Installed BiocManager and DESeq2

# 2: Import countData and colData
```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <- read.csv("airway_metadata.csv", stringsAsFactors = FALSE)
```

How many genes are we looking at?
```{r}
nrow(counts)
```

# 3: Toy differential gene expression
We want to calculate the average counts per gene for control samples and for the drug treated samples, and then compare those average to see if there are significant differences.
```{r}
#This gives the IDs that are the control samples in metadata, and stores it in a new data frame with all the attached information for each ID from metadata.
control <- metadata[metadata$dex == "control",]
```

```{r}
#Now we use $id of the control data frame to access the control columns in counts and calculate the average. 
control.mean <- rowSums(counts[ ,control$id]) / nrow(control)
names(control.mean) <- counts$ensgene
```

Now we do the same for treated.
```{r}
treated <- metadata[metadata$dex == "treated",]
treated.mean <- rowSums(counts[ ,treated$id]) / nrow(treated)
names(treated.mean) <- counts$ensgene
```
Now we create a new data frame where we combine these averages together.
```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

Now we plot control averages vs treated averages. If we see a straight line, it tells us that the drug has little to no effect.

```{r}
plot(meancounts, log = "xy")
```

We can then single out genes that have a large change between control and treated samples. We can see the log2 of the fold change.
```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```
There are weird results: Inf and NaN. We have to remove these from the data set. We are just interested in the genes that have 0 values. 
```{r}
#We look at meancounts, and see which columns  (treated and control means) for each row have values of zero. 
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)
#here, we store the unique genes (we don't want repeats) in a new variable called to.rm, and then we make a new data.frame called mycounts where we subtract those values stored in to.rm.
to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
#how many genes do we have left?
nrow(mycounts)
```

A common threshold used for calling something differentially expressed is a log2 foldchange of greater than 2 or less than -2. Let's filter the dataset both ways to see how many genes are up or down-regulated.
```{r}
#We store the values that have a fold change greater than 2 in up.ind, and the values that have a fold change less than -2 in down.ind. 
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

# 4: Adding annotation data

# 5: DESeq2 Analysis
Set up input for DESeq by specifying countdata, metadata, and the experiment design (the comparisons we are interested in).
```{r}
library(DESeq2)
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```

```{r}
dds <- DESeq(dds)
```
```{r}
res <- results(dds)
res
```

Padj is the adjusted p-value (we have to make it more stringent in order because we have so many genes to worry about - if the p-value threshold were 0.05, we would have a ton of false-positives).
We then plot the fold change vs the padj. We take the log of padj in order to get a linear trend that is easier to visualize. We take the negative so that it isn't upside down.
```{r}
plot(res$log2FoldChange, -log(res$padj))
```

```{r}
mycols <- rep("gray", nrow(res))
mycols[res$log2FoldChange > 2] <- "blue"
mycols[res$log2FoldChange < -2] <- "red"
plot(res$log2FoldChange, -log(res$padj), col = mycols)
abline(v=c(2,-2))
abline(h=-log(0.05))
```









